include/Dialect/Secret/IR/SecretOps.td

```diff
@@ -0,0 +1,161 @@
+#ifndef HEIR_INCLUDE_DIALECT_SECRET_IR_SECRETOPS_TD_
+#define HEIR_INCLUDE_DIALECT_SECRET_IR_SECRETOPS_TD_
+
+include "SecretDialect.td"
+include "SecretTypes.td"
+include "mlir/IR/BuiltinAttributeInterfaces.td"
+include "mlir/Interfaces/ControlFlowInterfaces.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class Secret_Op<string mnemonic, list<Trait> traits = []> :
+        Op<Secret_Dialect, mnemonic, traits> {
+  let cppNamespace = "::mlir::heir::secret";
+}
+
+def Secret_ConcealOp : Secret_Op<"conceal", [Pure]> {
+  let summary = "Convert a non-secret value into a secret";
+  let description = [{
+    Convert a value to a secret containing the same value.
+
+    This op represents a scheme-agnostic encryption operation, as well as a
+    "trivial encryption" operation which is needed for some FHE schemes. This
+    op is also useful for type materialization in the dialect conversion
+    framework.
+
+    Examples:
+
+    ```
+    %Y = secret.conceal %value : i32 -> !secret.secret<i32>
+    ```
+  }];
+
+  let arguments = (ins AnyType:$cleartext);
+  let results = (outs Secret:$output);
+  let assemblyFormat = "$cleartext attr-dict `:` type($cleartext) `->` type($output)";
+
+  let builders = [
+    // Builder to infer output type from the input type
+    OpBuilder<(ins "Value":$cleartext)>
+  ];
+
+  // TODO: add a folder that collapses secret<secret<T>> into secret<T>
+  // and removes intermediate conceal ops.
+  // let hasFolder = 1;
+}
+
+def Secret_RevealOp : Secret_Op<"reveal", [Pure]> {
+  let summary = "Convert a secret value into a non-secret";
+  let description = [{
+    Convert a secret into a non-secret containing the same value.
+
+    This op represents a scheme-agnostic decryption operation. This op is also
+    useful for target materialization in the dialect conversion framework.
+
+    Examples:
+
+    ```
+    %Y = secret.reveal %secret_value : !secret.secret<i32> -> i32
+    ```
+  }];
+
+  let arguments = (ins Secret:$input);
+  let results = (outs AnyType:$cleartext);
+  let assemblyFormat = "$input attr-dict `:` type($input) `->` type($cleartext)";
+
+  let builders = [
+    // Builder to infer output type from the input type
+    OpBuilder<(ins "Value":$secret)>
+  ];
+}
+
+
+def Secret_YieldOp : Secret_Op<"yield", [Pure, ReturnLike, Terminator]>,
+  Arguments<(ins Variadic<AnyType>:$values)> {
+  let summary = "Secret yield operation";
+  let description = [{
+    `secret.yield` is a special terminator operation for blocks inside regions
+    in `secret` generic ops. It returns the cleartext value of the
+    corresponding private computation to the immediately enclosing `secret`
+    generic op.
+  }];
+  let builders = [OpBuilder<(ins), [{ /* nothing to do */ }]>];
+  let hasCustomAssemblyFormat = 1;
```

Comment:
this might be a candidate for a verifier (i remember when we were pair programming)

Reply:
Added to the TODOs and a note about SingleBlockImplicitTerminator